File: ios/Runner/Routine.swift
//
//  Routine.swift
//  Runner
//
//  Created by Ajan on 2/28/25.
//

import ManagedSettings
import Foundation
import os.log

class Routine: Codable {
    let id: String
    let name: String
    let days: [Bool]
    let startTime: Int?
    let endTime: Int?
    let allDay: Bool
    let pausedUntil: Date?
    let snoozedUntil: Date?
    let conditionsLastMet: Date?
    let strictMode: Bool?
    private(set) var apps: [ApplicationToken]
    private(set) var sites: [WebDomainToken]
    private(set) var domains: [String]
    private(set) var categories: [ActivityCategoryToken]
    
    enum CodingKeys: String, CodingKey {
        case id, name, days, startTime, endTime, allDay, pausedUntil, snoozedUntil, apps, sites, categories, allow, conditionsMet, conditionsLastMet, strictMode
    }
    let allow: Bool
    
    // Required initializer for Decodable protocol
    required init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        
        self.id = try container.decode(String.self, forKey: .id)
        self.name = try container.decode(String.self, forKey: .name)
        self.days = try container.decode([Bool].self, forKey: .days)
        self.startTime = try container.decodeIfPresent(Int.self, forKey: .startTime)
        self.endTime = try container.decodeIfPresent(Int.self, forKey: .endTime)
        self.allDay = try container.decode(Bool.self, forKey: .allDay)
        self.allow = try container.decode(Bool.self, forKey: .allow)
        self.strictMode = try container.decodeIfPresent(Bool.self, forKey: .strictMode)

        // Handle conditionsLastMet as an optional Date
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        
        if let conditionsLastMetString = try container.decodeIfPresent(String.self, forKey: .conditionsLastMet) {
            self.conditionsLastMet = formatter.date(from: conditionsLastMetString)
        } else {
            self.conditionsLastMet = nil
        }

        // Handle optional Date properties
        
        if let pausedUntilString = try container.decodeIfPresent(String.self, forKey: .pausedUntil) {
            self.pausedUntil = formatter.date(from: pausedUntilString)
        } else {
            self.pausedUntil = nil
        }

        if let snoozedUntilString = try container.decodeIfPresent(String.self, forKey: .snoozedUntil) {
            self.snoozedUntil = formatter.date(from: snoozedUntilString)
        } else {
            self.snoozedUntil = nil
        }
        
        // Process tokens directly
        self.apps = [ApplicationToken]()
        self.sites = [WebDomainToken]()
        self.categories = [ActivityCategoryToken]()
        self.domains = [String]()
        
        // Since the tokens are stored as JSON strings, we need to decode them separately
        // We'll use a single JSONDecoder instance for all token types
        let jsonDecoder = JSONDecoder()
        
        if let appsData = try container.decodeIfPresent([String].self, forKey: .apps) {
            for appString in appsData {
                if let data = appString.data(using: .utf8),
                   let token = try? jsonDecoder.decode(ApplicationToken.self, from: data) {
                    self.apps.append(token)
                } else {
                    print("failed to create app token")
                }
            }
        }
        
        if let sitesData = try container.decodeIfPresent([String].self, forKey: .sites) {
            for siteString in sitesData {
                if let data = siteString.data(using: .utf8),
                   let token = try? jsonDecoder.decode(WebDomainToken.self, from: data) {
                    self.sites.append(token)
                } else {
                    self.domains.append(siteString)
                }
            }
        }
        
        if let categoriesData = try container.decodeIfPresent([String].self, forKey: .categories) {
            for categoryString in categoriesData {
                if let data = categoryString.data(using: .utf8),
                   let token = try? jsonDecoder.decode(ActivityCategoryToken.self, from: data) {
                    self.categories.append(token)
                } else {
                    print("failed to create category token")
                }
            }
        }
    }
    
    func isActive() -> Bool {
        let now = Date().addingTimeInterval(45)
        let calendar = Calendar.current
        
        // Get day of week (0-6, where 0 is Sunday in Swift)
        // Convert to 0-6 where 0 is Monday to match Dart implementation
        var dayOfWeek = calendar.component(.weekday, from: now) - 1 // 1-7 (Sunday is 1) -> 0-6
        dayOfWeek = (dayOfWeek + 6) % 7 // Convert to 0-6 where 0 is Monday
        
        // Check if routine is snoozed
        if let snoozedUntil = snoozedUntil, now < snoozedUntil {
            return false
        }
        
        // Check if routine is paused
        if let pausedUntil = pausedUntil, now < pausedUntil {
            return false
        }
        
        // If allDay or no time restrictions, just check the day
        if allDay || (startTime == nil && endTime == nil) {
            return dayOfWeek < days.count && days[dayOfWeek]
        }
        
        // Get current time in minutes since midnight
        let components = calendar.dateComponents([.hour, .minute], from: now)
        let currMins = (components.hour ?? 0) * 60 + (components.minute ?? 0)
        
        // Unwrap optional startTime and endTime with defaults
        if (startTime == nil || endTime == nil) {
            return true;
        }
        
        let start = startTime ?? -1
        let end = endTime ?? -1
        
        // If start time is after end time (crosses midnight)
        if end < start {
            if currMins >= start {
                // Current time is after start time but before midnight
                // Only need to check if current day is enabled
                return dayOfWeek < days.count && days[dayOfWeek]
            } else if currMins < end {
                // Current time is after midnight but before end time
                // Check if yesterday was enabled (routine started yesterday)
                let yesterdayOfWeek = (dayOfWeek + 6) % 7 // Previous day, wrapping from 0 back to 6
                return yesterdayOfWeek < days.count && days[yesterdayOfWeek]
            }
            return false
        }
        
        // Normal case: start time is before end time
        return dayOfWeek < days.count && days[dayOfWeek] && (currMins >= start && currMins < end)
    }
    
    func areConditionsMet() -> Bool {
        // If conditionsLastMet is nil, return false
        guard let lastMet = conditionsLastMet else {
            return false
        }
        
        // Get the current date and extract the time components
        let now = Date()
        let calendar = Calendar.current
        
        // Handle all-day routines
        if allDay || (startTime == nil && endTime == nil) {
            // For all-day routines, just check if conditionsLastMet is today
            return calendar.isDateInToday(lastMet)
        }
        
        // Unwrap optional startTime and endTime with defaults
        guard let start = startTime, let end = endTime else {
            return false
        }
        
        // Get current time components
        let nowComponents = calendar.dateComponents([.year, .month, .day, .hour, .minute], from: now)
        let currDate = calendar.date(from: nowComponents) ?? now
        
        // Get today's date at start time
        var startComponents = calendar.dateComponents([.year, .month, .day], from: now)
        startComponents.hour = start / 60
        startComponents.minute = start % 60
        startComponents.second = 0
        guard let todayAtStartTime = calendar.date(from: startComponents) else {
            return false
        }
        
        // If this is an overnight routine (start time > end time)
        if end < start {
            // Get yesterday's date at start time
            guard let yesterdayDate = calendar.date(byAdding: .day, value: -1, to: currDate) else {
                return false
            }
            var yesterdayComponents = calendar.dateComponents([.year, .month, .day], from: yesterdayDate)
            yesterdayComponents.hour = start / 60
            yesterdayComponents.minute = start % 60
            yesterdayComponents.second = 0
            
            guard let yesterdayAtStartTime = calendar.date(from: yesterdayComponents) else {
                return false
            }
            
            // Get today's date at end time
            var endComponents = calendar.dateComponents([.year, .month, .day], from: now)
            endComponents.hour = end / 60
            endComponents.minute = end % 60
            endComponents.second = 0
            guard let todayAtEndTime = calendar.date(from: endComponents) else {
                return false
            }
            
            // Current time is after midnight but before end time
            if now < todayAtEndTime {
                // Check if conditions were met after yesterday's start time
                return yesterdayAtStartTime.compare(lastMet) == .orderedAscending
            }
            // Current time is after start time but before midnight
            else if now >= todayAtStartTime {
                // Check if conditions were met after today's start time
                return todayAtStartTime.compare(lastMet) == .orderedAscending
            }
            return false
        }
        
        // Normal case: start time is before end time
        // Return true if conditionsLastMet is after today's start time (completed during routine)
        return todayAtStartTime.compare(lastMet) == .orderedAscending
    }
}

// MARK: - Encodable Implementation
extension Routine {
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        
        try container.encode(id, forKey: .id)
        try container.encode(name, forKey: .name)
        try container.encode(days, forKey: .days)
        try container.encode(startTime, forKey: .startTime)
        try container.encode(endTime, forKey: .endTime)
        try container.encode(allDay, forKey: .allDay)
        try container.encode(allow, forKey: .allow)
        try container.encode(strictMode, forKey: .strictMode)

        // Create a single formatter for date properties
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        
        // Handle conditionsLastMet
        if let conditionsLastMet = conditionsLastMet {
            try container.encode(formatter.string(from: conditionsLastMet), forKey: .conditionsLastMet)
        } else {
            try container.encodeNil(forKey: .conditionsLastMet)
        }
        
        // Handle optional Date properties
        if let pausedUntil = pausedUntil {
            try container.encode(formatter.string(from: pausedUntil), forKey: .pausedUntil)
        } else {
            try container.encodeNil(forKey: .pausedUntil)
        }
        
        if let snoozedUntil = snoozedUntil {
            try container.encode(formatter.string(from: snoozedUntil), forKey: .snoozedUntil)
        } else {
            try container.encodeNil(forKey: .snoozedUntil)
        }
        
        // Create a single JSONEncoder for token encoding
        let jsonEncoder = JSONEncoder()
        
        // Encode tokens as strings
        let appsData = try apps.map { try String(data: jsonEncoder.encode($0), encoding: .utf8) ?? "" }
        try container.encode(appsData, forKey: .apps)
        
        let sitesData = try sites.map { try String(data: jsonEncoder.encode($0), encoding: .utf8) ?? "" }
        try container.encode(sitesData, forKey: .sites)
        
        let categoriesData = try categories.map { try String(data: jsonEncoder.encode($0), encoding: .utf8) ?? "" }
        try container.encode(categoriesData, forKey: .categories)
    }
}

File: android/app/src/main/kotlin/com/solidsoft/routine/Routine.kt
package com.solidsoft.routine

import android.util.Log
import org.json.JSONObject
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone

class Routine(
    val id: String,
    val name: String,
    val days: List<Boolean>,
    val startTime: Int?,
    val endTime: Int?,
    val allDay: Boolean,
    val pausedUntil: Date?,
    val snoozedUntil: Date?,
    val conditionsLastMet: Date?,
    val allow: Boolean,
    val strictMode: Boolean = false,
    private val apps: MutableList<String> = mutableListOf(),
    private val sites: MutableList<String> = mutableListOf()
) {
    companion object {
        private const val TAG = "Routine"
        
        // JSON keys matching Swift implementation
        const val KEY_ID = "id"
        const val KEY_NAME = "name"
        const val KEY_DAYS = "days"
        const val KEY_START_TIME = "startTime"
        const val KEY_END_TIME = "endTime"
        const val KEY_ALL_DAY = "allDay"
        const val KEY_PAUSED_UNTIL = "pausedUntil"
        const val KEY_SNOOZED_UNTIL = "snoozedUntil"
        const val KEY_CONDITIONS_LAST_MET = "conditionsLastMet"
        const val KEY_STRICT_MODE = "strictMode"
        const val KEY_ALLOW = "allow"
        const val KEY_APPS = "apps"
        const val KEY_SITES = "sites"
        const val KEY_CATEGORIES = "categories"
        
        // Date formatter matching Swift's ISO8601DateFormatter with fractional seconds
        val iso8601Formatter = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSSSS'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }
        
        fun fromJson(jsonString: String): Routine? {
            return try {
                val json = JSONObject(jsonString)
                Routine(json)
            } catch (e: Exception) {
                Log.e(TAG, "Error parsing routine JSON: ${e.message}")
                null
            }
        }
    }
    
    /**
     * Secondary constructor that creates a Routine from a JSONObject
     */
    constructor(json: JSONObject) : this(
        id = json.getString(KEY_ID),
        name = json.getString(KEY_NAME),
        days = List(json.getJSONArray(KEY_DAYS).length()) { i -> json.getJSONArray(KEY_DAYS).getBoolean(i) },
        startTime = if (json.has(KEY_START_TIME) && !json.isNull(KEY_START_TIME)) json.getInt(KEY_START_TIME) else null,
        endTime = if (json.has(KEY_END_TIME) && !json.isNull(KEY_END_TIME)) json.getInt(KEY_END_TIME) else null,
        allDay = json.getBoolean(KEY_ALL_DAY),
        pausedUntil = if (json.has(KEY_PAUSED_UNTIL) && !json.isNull(KEY_PAUSED_UNTIL)) 
            iso8601Formatter.parse(json.getString(KEY_PAUSED_UNTIL)) else null,
        snoozedUntil = if (json.has(KEY_SNOOZED_UNTIL) && !json.isNull(KEY_SNOOZED_UNTIL)) 
            iso8601Formatter.parse(json.getString(KEY_SNOOZED_UNTIL)) else null,
        conditionsLastMet = if (json.has(KEY_CONDITIONS_LAST_MET) && !json.isNull(KEY_CONDITIONS_LAST_MET)) 
            iso8601Formatter.parse(json.getString(KEY_CONDITIONS_LAST_MET)) else null,
        allow = json.getBoolean(KEY_ALLOW),
        strictMode = if (json.has(KEY_STRICT_MODE) && !json.isNull(KEY_STRICT_MODE)) json.getBoolean(KEY_STRICT_MODE) else false
    ) {
        // Parse apps array
        if (json.has(KEY_APPS) && !json.isNull(KEY_APPS)) {
            val appsArray = json.getJSONArray(KEY_APPS)
            for (i in 0 until appsArray.length()) {
                apps.add(appsArray.getString(i))
            }
        }

        if (json.has(KEY_SITES) && !json.isNull(KEY_SITES)) {
            val sitesArray = json.getJSONArray(KEY_SITES)
            for (i in 0 until sitesArray.length()) {
                sites.add(sitesArray.getString(i))
            }
        }
    }

    fun isActive(): Boolean {
        // Add 45 seconds buffer to match Swift implementation
        val now = Date(System.currentTimeMillis() + 45000)
        val calendar = Calendar.getInstance()
        calendar.time = now
        
        // Get day of week (0-6, where 0 is Monday to match Swift implementation)
        var dayOfWeek = (calendar.get(Calendar.DAY_OF_WEEK) + 5) % 7 // Convert from Calendar.DAY_OF_WEEK (1-7, Sunday is 1) to 0-6 where 0 is Monday
        
        // Check if routine is snoozed
        if (snoozedUntil != null && now.before(snoozedUntil)) {
            return false
        }
        
        // Check if routine is paused
        if (pausedUntil != null && now.before(pausedUntil)) {
            val dateFormatter = SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US)
            val dateString = dateFormatter.format(pausedUntil)
            val nowString = dateFormatter.format(now)
            Log.d(TAG, "Routine is paused - $nowString < $dateString")
            return false
        }
        
        // If allDay or no time restrictions, just check the day
        if (allDay || (startTime == null && endTime == null)) {
            return dayOfWeek < days.size && days[dayOfWeek]
        }
        
        // Get current time in minutes since midnight
        val currMins = calendar.get(Calendar.HOUR_OF_DAY) * 60 + calendar.get(Calendar.MINUTE)
        
        // Unwrap optional startTime and endTime with defaults
        if (startTime == null || endTime == null) {
            return true
        }
        
        val start = startTime ?: -1
        val end = endTime ?: -1
        
        // If start time is after end time (crosses midnight)
        if (end < start) {
            if (currMins >= start) {
                // Current time is after start time but before midnight
                // Only need to check if current day is enabled
                return dayOfWeek < days.size && days[dayOfWeek]
            } else if (currMins < end) {
                // Current time is after midnight but before end time
                // Check if yesterday was enabled (routine started yesterday)
                val yesterdayOfWeek = (dayOfWeek + 6) % 7 // Previous day, wrapping from 0 back to 6
                return yesterdayOfWeek < days.size && days[yesterdayOfWeek]
            }
            return false
        }
        
        // Normal case: start time is before end time
        return dayOfWeek < days.size && days[dayOfWeek] && (currMins >= start && currMins < end)
    }
    
    fun areConditionsMet(): Boolean {
        // If conditionsLastMet is null, return false
        val lastMet = conditionsLastMet ?: return false
        
        // Get the current date and extract the time components
        val now = Date()
        val calendar = Calendar.getInstance()
        
        // Reset calendar to today at midnight
        calendar.time = now
        calendar.set(Calendar.HOUR_OF_DAY, 0)
        calendar.set(Calendar.MINUTE, 0)
        calendar.set(Calendar.SECOND, 0)
        calendar.set(Calendar.MILLISECOND, 0)
        
        // Use startTime if defined, otherwise use 0 (midnight)
        val effectiveStartTime = startTime ?: 0
        
        val startHour = effectiveStartTime / 60
        val startMinute = effectiveStartTime % 60
        
        // Set calendar to today at start time
        calendar.set(Calendar.HOUR_OF_DAY, startHour)
        calendar.set(Calendar.MINUTE, startMinute)
        
        val todayAtStartTime = calendar.time
        
        // Return true if conditionsLastMet is after today's start time (completed during routine)
        return lastMet.after(todayAtStartTime)
    }
    
    // Getter methods for collections
    fun getApps(): List<String> = apps.toList()
    
    fun getSites(): List<String> = sites.toList()
}
